### PR TITLE
Line thickness

### PR DESCRIPTION
**Pull request by fsih**
### Resolves
Fixes
https://github.com/LLK/scratch-paint/issues/618

Before
image
After
quadraticmeme

### Proposed Changes
Preprocess SVGS:
- Push transforms down to the leaf level, composing as we go
- Push stroke widths down to the leaf level. Stroke widths overwrite each other.
- At the leaf level, if
    - The node is a path
    - with transform
    - with a stroke width
then apply the transform to the path (replacing the path data), thereby removing the transform. Then apply the scale factors of the transform to the stroke width, and give the entire path stroke width equal to the quadratic mean of the change in percent of stroke width x and stroke width y times the original stroke width.

### Reason for Changes
That's what Flash (and therefore Scratch 2) does when it parses paths with stroke widths for display. (Transformed paths with stroke widths change stroke width, but the stroke width stays even all the way around, unlike the default SVG behavior.) This way stroke widths on transformed shapes will look like they do in Scratch 2.

### Test Coverage
I wrote tests that generate an HTML file for visually comparing the preprocessed SVG with the original SVG with transform.

### CODE REVIEW

**Comment by fsih** [comment]
If you look at the test output file transform-applier-test.html, you can see there's still some issue with arbitrary transforms on elliptical arcs, but I figured I should get started on the review process.

Elliptical arc curves are the only SVG instruction that doesn't consist of coordinates. They are defined by an ellipse (2 radii and a rotation angle) plus 2 points that the ellipse passes through, which only narrows it down to 4 possible arcs, so plus 2 more parameters that define which of the 4 arcs to choose. If the ellipse can't reach both points, it is scaled up until it can. They're very troublesome to transform =_=

If I don't figure it out maybe I'll use https://www.npmjs.com/package/svg-arc-to-cubic-bezier. Bezier curves can't exactly define elliptical arcs, but it's hard to tell the difference.

**Comment by fsih** [comment]
Travis test failing because one number differs in the 14th decimal place
Floating point errors differ on different machines

(ノ°Д°）ノ︵ ┻━┻

**Comment by cwillisf** [comment]
Just to make sure I understand the goal here: are these statements accurate?

1. A standards-compliant SVG can have a transform that affects a whole branch of the SVG element tree
2. Paper.js and/or our Scratch 3.0 SVG rendering code don't support such transforms
3. The only real workaround is to "apply" the transforms to the affected elements such that the resulting tree is visually the same but contains no transform nodes

So basically turn `<transform="move right 1"><transform="move right 1"><cat x=1 y=2 /></transform></transform>` into `<cat x=3 y=2 />`?

If that's the case: have you looked to see if there's already a library out there which does this? It's awesome that you're working through all this math but I worry that you're going to keep finding more attributes which need to be transformed in even stranger ways...

**Comment by fsih** [comment]
No, the transforms are all working fine, the issue I'm trying to fix is the stroke width difference between Scratch 2.0 and 3.0. There's 2 differently-presenting issues here, one on stage and one in the paint editor.

On stage:
The issue can be seen in this project: https://llk.github.io/scratch-gui/develop/#203392290
2.0:
image
3.0: 
image
What's happening is that this is made up of a bunch of paths like this, that are scaled unevenly in X and Y
```
<svg version="1.1" width="865" height="761" viewBox="0 -1 865 761" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
  <path id="ID0.7168996930122375" fill="none" stroke="#6CBEFF" stroke-width="15" stroke-linecap="round" d="M 116.400 178.750 L 223.650 99 " transform="matrix(0.8831499814987183, 0.4690907299518585, -1.447621464729309, 2.725414991378784, 458.6, 199.6)"/>
</svg>
```
This is how the browser renders such a path. The renderer is doing exactly the right thing, according to how SVG is defined.
image
However, what Scratch 2 does is it parses SVG path commands into an AS3 Shape object, and then applies the transform and stroke width properties to that Shape. Actionscript treats stroke widths on transformed Shapes differently from SVG. It basically draws the path first, and then applies an averaged out stroke width to the whole shape, so the stroke width is even all around. You can see that stretching in the X direction thickens the top and bottom strokes:
strokewidthsb2
So this is in the class of issues where Scratch 2 is exporting an SVG that isn't the same as what it's displaying.
I'm going to make a new comment since this one is getting too long.

**Comment by fsih** [comment]
In the paint editor:
Paper.js actually behaves more like ActionScript. When you import an SVG, it applies the transform to all the path coordinates first, and then draws a stroke on that. So Paper is doing part of what I'm doing in this PR, which is why I was able to borrow a lot of code from Paper.

The advantage of this method is that all your shapes aren't in different coordinate spaces. Any path that says it has a stroke width of 3 has a stroke 3 times thicker than any path that says it has a stroke width of 1 (except the text tool...). And the stroke width will be 3 all the way around. On the other hand, if Scratch 2 showed a number for the stroke width, it might not be very useful, because scaling the shape wouldn't affect that number.

The problem is that Paper doesn't also do the step where it averages out the stroke width while doing the applyMatrix step. It just leaves the stroke width as whatever its original unscaled size was. Here's that ICE background in the paint editor.
image
In fact, since the transforms are removed by the applyMatrix step, the information about how the stroke width was transformed is lost, so it can't be fixed after the fact.

One solution would be to turn off applyMatrix. That would make the imported SVG look like it does on stage, but it would also make coordinate systems inconsistent, and all the tools that made assumptions about the coordinate system would have to be fixed.
https://github.com/paperjs/paper.js/blob/a61e83edf2ed1870bd41bad135f4f6fc85b0f628/examples/Paperjs.org/PathIntersections.html#L9

The solution that I'm going with is figuring out what the transform on the stroke width looks like before the applyMatrix step happens, and then transforming the stroke widths. I could do that in Paper.js, but I decided to do it in SvgRenderer because stepping through the whole SVG tree is necessary anyway to fix the issue that we see on the stage.

I agree that this change feels risky and it would be great to rely on a library, but I'm not sure how to find one since it's pretty specific to how Scratch 2 worked. There's not a "right way" to normalize stroke widths. (In fact, I found the way AS3 does it (quadratic mean) pretty unintuitive, and it doesn't even match Adobe's other product Illustrator, which uses a geometric mean.) But since it's affecting enough projects, it's considered launch-blocking, so I think the best I can do is try to get it in asap so we have time to find anything that breaks :/

It's hard to clarify in text, I'm happy to video chat whenever if you want!

**Comment by fsih** [comment]
There's still an issue with importing shapes with a gradient with a skew component, but putting this back up for review because it's more important that we test what we have than that we fix that last thing right now :/

File: src/transform-applier.js [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+ * @param {SVGElement} svgTag The SVG dom object
+ * @param {Window} windowRef The window to use. Need to pass in for
+ *     tests to work, as they get angry at even the mention of window.
+ * @param {object} bboxForTesting The bounds to use. Need to pass in for
+ *     tests only, because getBBox doesn't work in Node. This should
+ *     be the bounds of the svgTag without including stroke width or transforms.
+ * @return {void}
+ */
+const transformStrokeWidths = function (svgTag, windowRef, bboxForTesting) {
+    // TODO x2, r, cy, cx, fy, fx defaults are wrong.
```
Comment (fsih) [inline src/transform-applier.js]: Take off todo

File: src/transform-applier.js [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+            }
+
+            // Transform path data
+            element.setAttribute('d', _transformPath(element.attributes.d.value, matrix));
+            element.removeAttribute('transform');
+
+            // Transform stroke width
+            const matrixScale = _getScaleFactor(matrix);
+            element.setAttribute('stroke-width', _quadraticMean(matrixScale.x, matrixScale.y) * strokeWidth);
+            element.setAttribute('fill', fill);
```
Comment (fsih) [inline src/transform-applier.js:542]: Test what happens here when you set fill to undefined

File: src/transform-applier.js [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+    // First split the path data into parts of command-coordinates pairs
+    // Commands are any of these characters: mzlhvcsqta
+    const parts = pathString && pathString.match(/[mlhvcsqtaz][^mlhvcsqtaz]*/ig);
+    let coords;
+    let relative = false;
+    let previous;
+    let control;
+    let current = {x: 0, y: 0};
+    let start = {x: 0, y: 0};
+    let translated = '';
```
Comment (fsih) [inline src/transform-applier.js]: Rename to result?

File: src/transform-applier.js [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+    const b = -transform.c;
+    const c = -transform.b;
+    const d = transform.d;
+    // Since other parameters determine the translation of the ellipse in SVG, we do not need to worry
+    // about what e and f are.
+    const det = (a * d) - (b * c);
+    // Non-invertible matrix
+    if (det === 0) return null;
+
+    // rotA, rotB, and rotC represent Ax^2 + Bxy + Cy^2 = 1 coefficients for a rotated ellipse formula
```
Comment (fsih) [inline src/transform-applier.js:89]: Consider storing cos theta, sin theta, sin 2 theta in variables to not have to repeat calculation

File: src/transform-applier.js [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+// Adapted from paper.js's SvgImport.getDefinition
+const _parseUrl = (value, windowRef) => {
+    // When url() comes from a style property, '#'' seems to be missing on
+    // WebKit. We also get variations of quotes or no quotes, single or
+    // double, so handle it all with one regular expression:
+    const match = value && value.match(/\((?:["'#]*)([^"')]+)/);
+    const name = match && match[1];
+    const res = name && windowRef ?
+        // This is required by Firefox, which can produce absolute
+        // urls for local gradients, see #1001:
```
Comment (fsih) [inline src/transform-applier.js]: Write that this is a paper issue number

File: src/transform-applier.js [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+    return strokeWidth &&
+        element.tagName && element.tagName.toLowerCase() === 'path' &&
+        element.attributes.d && element.attributes.d.value;
+};
+const _quadraticMean = function (a, b) {
+    return Math.sqrt(((a * a) + (b * b)) / 2);
+};
+
+const _createGradient = function (gradientId, svgTag, bbox, matrix) {
+    const getValue = function (node, name, isString, allowNull, allowPercent, defaultValue) {
```
Comment (fsih) [inline src/transform-applier.js:313]: Write that this is from Paper

File: src/transform-applier.js (hunk 2 above)
Comment (fsih) [reply src/transform-applier.js:542]: I tried it out, and it makes the shape fill with black.

https://www.w3.org/TR/SVG11/painting.html#FillProperties
It looks like that's correct according to the spec.

**Review by cwillisf — APPROVED** [review]
Awesome -- thanks for all the hard work!
😅👍

**Comment by fsih** [comment]
Oh boy, here goes...
